packages/SwingSet/src/kernel/vat-warehouse.js

```diff
@@ -3,11 +3,142 @@ import { assert, Fail, quote as q } from '@agoric/assert';
 import { isNat } from '@endo/nat';
 import { makeVatTranslators } from './vatTranslator.js';
 import { insistVatDeliveryResult } from '../lib/message.js';
+import djson from '../lib/djson.js';
 
 /**
  * @typedef {import('@agoric/swingset-liveslots').VatDeliveryObject} VatDeliveryObject
+ * @typedef {import('@agoric/swingset-liveslots').VatDeliveryResult} VatDeliveryResult
+ * @typedef {import('@agoric/swingset-liveslots').VatSyscallObject} VatSyscallObject
+ * @typedef {import('@agoric/swingset-liveslots').VatSyscallResult} VatSyscallResult
+ * @typedef {import('@agoric/swingset-liveslots').VatSyscallHandler} VatSyscallHandler
  * @typedef {import('../types-internal.js').VatManager} VatManager
+ * @typedef {{ body: string, slots: unknown[] }} Capdata
+ * @typedef { [unknown, ...unknown[]] } Tagged
+ * @typedef { { moduleFormat: string }} Bundle
+ */
+
+/**
+ * @param {VatSyscallObject} originalSyscall
+ * @param {VatSyscallObject} newSyscall
+ * @returns {boolean}
+ */
+export function syscallsAreIdentical(originalSyscall, newSyscall) {
+  return djson.stringify(originalSyscall) === djson.stringify(newSyscall);
+}
+
+/**
+ * @param {VatSyscallHandler} origHandler
+ */
+function recordSyscalls(origHandler) {
+  assert(origHandler);
+  const syscalls = [];
+  const syscallHandler = vso => {
+    const vres = origHandler(vso);
+    syscalls.push({ s: vso, r: vres });
+    return vres;
+  };
+  const getSyscalls = () => syscalls;
+  return { syscallHandler, getSyscalls };
+}
+
+/**
+ * Make a syscallHandler that returns results from a
+ * previously-recorded transcript, instead of executing them for
+ * real. The vat must perform exactly the same syscalls as before,
+ * else it gets a vat-fatal error.
+ *
+ * @param {*} kernelSlog
+ * @param {string} vatID
+ * @param {number} deliveryNum
+ * @param {import('../types-external.js').TranscriptEntry} transcriptEntry
+ * @returns { {
+ *   syscallHandler: (vso: VatSyscallObject) => VatSyscallResult,
+ *   finishSimulation: () => void,
+ * } }
  */
+function makeSyscallSimulator(kernelSlog, vatID, deliveryNum, transcriptEntry) {
+  const syscallsExpected = Array.from(transcriptEntry.sc); // copy
```

Comment:
The colloquial way is usually to use spread.
```suggestion
  const syscallsExpected = [...transcriptEntry.sc]; // copy
```